### PR TITLE
C#: Use explicit native extension loading whenever possible

### PR DESCRIPTION
**Pull request by jtattermusch**
Tentatively fixes the hang issue described in https://github.com/grpc/grpc/issues/25208.

The problem is a side effect of https://github.com/grpc/grpc/pull/24744.

### CODE REVIEW

**Comment by jtattermusch** [comment]
CC @jskeet

**Comment by jtattermusch** [comment]
Adhoc package/distribtest build: https://fusion2.corp.google.com/invocations/8ea33ee1-a3d8-4d75-bc37-89071a20ab4e/targets

**Comment by jtattermusch** [comment]
One more adhoc run of packages/distribtests: https://fusion2.corp.google.com/invocations/03d594b8-1240-43e2-9cb3-b505e14318fc/targets

**Comment by jtattermusch** [comment]
Also fixes https://github.com/grpc/grpc/pull/7303

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs [hunk 1]
```diff
@@ -78,29 +80,80 @@ public NativeMethods NativeMethods
...
+            // (while if we used [DllImport], the native extension won't be on the other app's search path for shared libraries).
             var assemblyDirectory = GetAssemblyDirectory();
 
             // With "classic" VS projects, the native libraries get copied using a .targets rule to the build output folder
             // alongside the compiled assembly.
+            // With dotnet SDK projects targeting net45 framework, the native libraries (just the required ones)
+            // are similarly copied to the built output folder, through the magic of Microsoft.NETCore.Platforms.
             var classicPath = Path.Combine(assemblyDirectory, GetNativeLibraryFilename());
 
+            // With dotnet SDK project targeting netcoreappX.Y, projects will use Grpc.Core assembly directly in the location where it got restored
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs:99]: this just adds back code that was removed in https://github.com/grpc/grpc/pull/24744/files#diff-2c65a0fc0ff30ef55102682b95465c44e758e53921b0f36be1b427f1feb12e18L95

**Comment by jtattermusch** [comment]
The adhoc build of nugets is here:
https://storage.cloud.google.com/grpc-testing-kokoro-prod/test_result_public/prod/grpc/core/master/linux/grpc_build_packages/15732/20210218-065115/github/grpc/artifacts/csharp_nugets_windows_dotnetcli.zip

@jskeet do you want to test the nugets and see if that fixes the problems you've encountered?

All the distribtests seem to be passing.

**Comment by jskeet** [comment]
I'll do a test tomorrow, thanks.

**Review by apolcyn — APPROVED** [review]

**Comment by jskeet** [comment]
My specific "hang test" worked, as did the smoke tests for an actual API - great! Thanks very much :)

**Pull request merged**